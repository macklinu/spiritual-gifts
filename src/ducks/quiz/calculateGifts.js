// @ts-check

import { Map } from 'immutable'

export default answers =>
  answers.reduce(reduceAnswersToMap, Map()).sort(sortGreatestToLeast)

const mapColumnToGiftKey = column => String.fromCharCode(97 + column)

const reduceAnswersToMap = (map, answer, index) => {
  const column = index % 17
  const key = mapColumnToGiftKey(column)
  return map.update(key, value => (value || 0) + answer)
}

const sortGreatestToLeast = (a, b) => (a < b ? 1 : a > b ? -1 : 0)
