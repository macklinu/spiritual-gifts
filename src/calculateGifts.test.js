import { List } from 'immutable'
import calculateGifts from './calculateGifts'

test('calculates gifts', () => {
  // fill answers with seed data
  // set a few answers to greater answers to check sorting
  const answers = List(Array(17).fill(0)).set(9, 1).set(1, 2).set(15, 3)

  expect(calculateGifts(answers)).toMatchSnapshot()
})
