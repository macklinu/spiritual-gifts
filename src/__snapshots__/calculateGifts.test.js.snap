// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculates gifts 1`] = `
Immutable.OrderedMap {
  p: 3,
  b: 2,
  j: 1,
  a: 0,
  c: 0,
  d: 0,
  e: 0,
  f: 0,
  g: 0,
  h: 0,
  i: 0,
  k: 0,
  l: 0,
  m: 0,
  n: 0,
  o: 0,
  q: 0,
}
`;
